nixos/modules/services/misc/home-assistant.nix

```diff
@@ -0,0 +1,91 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.home-assistant;
+
+  configFile = pkgs.writeText "configuration.yaml" (builtins.toJSON cfg.config);
+in {
+  meta.maintainers = with maintainers; [ dotlambda ];
+
+  options.services.home-assistant = {
+    enable = mkEnableOption "Home Assistant";
+
+    configDir = mkOption {
+      default = "/var/lib/hass";
+      type = types.path;
+      description = "The config directory, where your <filename>configuration.yaml</filename> is located.";
+    };
+
+    config = mkOption {
+      default = null;
+      type = with types; nullOr attrs;
+      example = literalExample ''
+        {
+          homeassistant = {
+            name = "Home";
+            time_zone = "UTC";
+          };
+          frontend = { };
+          http = { };
+        }
+      '';
+      description = ''
+        Your <filename>configuration.yaml</filename> as a Nix attribute set.
+        Beware that setting this option will delete your previous <filename>configuration.yaml</filename>.
+      '';
+    };
+
+    package = mkOption {
+      default = pkgs.home-assistant;
+      defaultText = "pkgs.home-assistant";
+      type = types.package;
+      example = literalExample ''
+        pkgs.home-assistant.override {
+          extraPackages = ps: with ps; [ colorlog ];
+        }
+      '';
+      description = ''
+        Home Assistant package to use.
+
+        Most Home Assistant components require additional dependencies,
+        which are best specified by overriding <literal>pkgs.home-assistant</literal>.
+        You can find the dependencies by searching for failed imports in you log or by looking at this list:
+        <link xlink:href="https://github.com/home-assistant/home-assistant/blob/master/requirements_all.txt"/>
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.home-assistant = {
+      description = "Home Assistant";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      preStart = ''
+        # Make sure directories exist with correct owner
+        mkdir -p ${cfg.configDir}
+        chown -R hass:hass ${cfg.configDir}
+      '' + lib.optionalString (cfg.config != null) ''
+        rm -f ${cfg.configDir}/configuration.yaml
+        ln -s ${configFile} ${cfg.configDir}/configuration.yaml
+      '';
+      serviceConfig = {
+        PermissionsStartOnly = true; # preStart must be run as root
```

Comment:
Instead of setting ```PermissionsStartOnly``` and then having to manually ```mkdir``` and ```chown``` the directory, you could use one of the following options:

a) A ```tmpfiles.d``` fragment via ```systemd.tmpfiles.rules```
b) Set it as the home directory for the user and then have it created together with the user

While there is of course nothing technically wrong about your approach it's just nicer to use the declarative tools we have available.

And thanks for packaging this!!!

Reply:
I went for option b)

Reply:
Do you have any other comments regarding the module @peterhoeg? Or do you think it can be merged as is?

Reply:
The module looks very nice - and well done on the tests!!

Couple of smaller issues:

1. Please set ```serviceConfig.Restart = "on-failure";``` as we probably want it to restart if it crashes.

2. Have you looked at all at locking down the service even further with ```ProtectSystem```, ```PrivateTmp```, ```ReadOnlyDirectories``` and so on?

Reply:
1. Do we really want this?
I know Arch does it: https://aur.archlinux.org/cgit/aur.git/tree/home-assistant.service?h=home-assistant.
However, the official instructions don't have that: https://home-assistant.io/docs/autostart/systemd/.
What's the point of restarting if it fails? From my knowledge, home-assistant should only fail if there's something wrong with the config. Or do you have experience that suggests otherwise?
2. `PrivateTmp` can't be used because hass stores information on state and history inside `configDir`. `ReadOnlyDIrectories` is not possible for the same reason.
I would however implement `ProtectSystem=strict` and `ReadWritePaths=${configDir}`. Do you think that's an adequate choice?

Reply:
Is ```TMP``` inside ```configDir```?

Reply:
I thought you wanted to put `configDir` inside `/tmp`. Then it wouldn't make sense do delete `/tmp` at every restart. :)
With the current restrictions in place, does `PrivateTmp` have an additional benefit? It could even hinder some use cases, e.g. a certain [file sensor](https://home-assistant.io/components/sensor.file/).

Reply:
> I thought you wanted to put configDir inside /tmp. 

If home-assistant needs to persist data, it should do so under /var/lib as you already default it to.

> Then it wouldn't make sense do delete /tmp at every restart. :)

Well, technically PrivateTmp doesn't delete - it simply mounts a new directory every time it runs. But the effect is the same.

> With the current restrictions in place, does PrivateTmp have an additional benefit? It could even hinder some use cases, e.g. a certain file sensor.

If you need to access hass's /tmp from the outside, then it will not work. In terms of benefits, it's all about creating a clean, consistent environment in which to run to minimize the risk of something from the outside messing with it or broken/malicious code messing with things outside.

Reply:
Okay, I implemented it. I however had to move the test's `configDir` out of `/tmp`.